### PR TITLE
dap: add 'clipboard' support, and truncate a long value

### PR DESCRIPTION
**Pull request by hyangah**
* add 'clipboard' capability
* apply a larger string limit for 'hover' and 'clipboard' context
* truncate the string representation of a compound type (or pointer of a compound type) variable if the length of the value is larger than 256. But when a variable is returned as a result of Evaluate Request for hover or copy value feature (clipboard or variables context), do not truncate the value.

Updates golang/vscode-go#1318
Fixes golang/vscode-go#1435

### CODE REVIEW

**Comment by hyangah** [comment]
Arbitrarily chosen 1K, Still 1K is too large for my taste.

Before:
Screen Shot 2021-05-26 at 2 58 52 PM

After: 
Screen Shot 2021-05-26 at 2 59 42 PM

Still it's possible that we'd end up sending a large JSON message for variables response if the requested variable/scope contains many long variables. This PR doesn't mean to prevent that.

File: service/dap/server.go [hunk 1]
```diff
@@ -1613,26 +1614,34 @@ func (s *Server) getTypeIfSupported(v *proc.Variable) string {
...
+// type variable.
+const defaultMaxValueLen = 1 << 10 // 1K
+
 // convertVariableWithOpts allows to skip reference generation in case all we need is
-// a string representation of the variable.
-func (s *Server) convertVariableWithOpts(v *proc.Variable, qualifiedNameOrExpr string, skipRef bool) (value string, variablesReference int) {
+// a string representation of the variable. When the variable is a compound or reference
+// type variable and its full string representation can be larger than defaultMaxValueLen,
+// this returns a truncated value unless showFull is set.
+func (s *Server) convertVariableWithOpts(v *proc.Variable, qualifiedNameOrExpr string, skipRef, showFull bool) (value string, variablesReference int) {
```
Comment (aarzilli) [inline service/dap/server.go:1633]: I think this is enough boolean flags that it should be a bitfield.

**Comment by derekparker** [comment]
> Arbitrarily chosen 1K, Still 1K is too large for my taste.

Would you like time to live with it and possibly change this before it gets submitted?

**Comment by suzmue** [comment]
> Arbitrarily chosen 1K, Still 1K is too large for my taste.

I agree I think that these strings become unnecessarily long.

When I was testing paging, this was the cause of some things being so slow that it was unusable for investigating the values.

I would say drop it to ~100 (or even less) especially if it has children to expand to get more information.

File: service/dap/server.go (hunk 1 above)
Comment (hyangah) [reply service/dap/server.go]: done. PTAL

**Comment by hyangah** [comment]
Lowered the max limit to 256. Tried 128 and it broke some tests and some didn't look nice with 128 (e.g. slice of pointers to a compound type, ...) due to all the type info and len/cap info before the actual data. :-(
I think a better solution is to truncate more intelligently, but I don't want to go too deep into it now.

File: service/dap/server.go [hunk 2]
```diff
@@ -1814,26 +1815,40 @@ func (s *Server) getTypeIfSupported(v *proc.Variable) string {
...
+	val, _ := s.convertVariableWithOpts(v, "", skipRef)
 	return val
 }
 
+// defaultMaxValueLen is the max length of a string representation of a compound or reference
+// type variable.
+const defaultMaxValueLen = 1 << 8 // 256
+
+// Flags for convertVariableWithOpts option.
+const (
```
Comment (aarzilli) [inline service/dap/server.go:1831]: I'd make these of a custom type: `type convertVariableFlags uint8`

**Comment by hyangah** [comment]
@polinasok is this 256 limit ok? Or should we look into further to do better than SinglelineString and avoid truncating informational messages (e.g. failed load, or partial load...)?

**Review by aarzilli — APPROVED** [review]
LGTM, needs decision over the length constant.

**Comment by suzmue** [comment]
LGTM

**Review by derekparker — APPROVED** [review]
LGTM

I'll wait to merge until the length is sorted out and we get an ack from @polinasok.

File: service/dap/server.go [hunk 3]
```diff
@@ -2019,7 +2040,8 @@ func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
 			return
 		}
 
-		if exprVar.Kind == reflect.String && (request.Arguments.Context == "repl" || request.Arguments.Context == "variables") {
+		ctxt := request.Arguments.Context
+		if exprVar.Kind == reflect.String && (ctxt == "repl" || ctxt == "variables" || ctxt == "hover" || ctxt == "clipboard") {
```
Comment (polinasok) [inline service/dap/server.go:2044]: switch-case with multiple values for context might be easier to read here

**Review by polinasok — COMMENTED** [review]
Could you remind me please what actions in the vscode UI generate evaluate requests with "hover" and "clipboard" context. I know that right click + "Copy to Value" is "variabels" and using Debug Console is "repl".

File: service/dap/server.go [hunk 4]
```diff
@@ -2031,9 +2053,11 @@ func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
 				}
 			}
 		}
-		// TODO(polina): as far as I can tell, evaluateName is ignored by vscode for expression variables.
-		// Should it be skipped altogether for all levels?
-		exprVal, exprRef := s.convertVariable(exprVar, fmt.Sprintf("(%s)", request.Arguments.Expression))
+		var opts convertVariableFlags
+		if ctxt == "variables" || ctxt == "hover" || ctxt == "clipboard" {
```
Comment (polinasok) [inline service/dap/server.go:2057]: Sadly "Copy Value" from WATCH appears to result in "watch" context request.

**Review by polinasok — COMMENTED** [review]
The PR description needs updating.

**Comment by suzmue** [comment]
When the server declares the capability `supportsClipboardContext`, `Copy Value` from both the variable pane and WATCH use the clipboard context.

File: service/dap/server.go (hunk 4 above)
Comment (suzmue) [reply service/dap/server.go:2057]: Are you still getting this when running a dap server with supportsClipboardContext?

**Comment by hyangah** [comment]
@polinasok It looks like you accidentally edited [my comment](https://github.com/go-delve/delve/pull/2513#issuecomment-850876556) instead of responding to it.

> I am not sure what you mean by "truncating information messages".

- information other than the variable's value such as len/cap of slice/array, or type information or address expression
- information that indicates the loading states, such as " ... 2 more", "- FAILED TO LOAD: some error", ...

> Could you remind me please what actions in the vscode UI generate evaluate requests with "hover" and "clipboard" context. I know that right click + "Copy to Value" is "variabels" and using Debug Console is "repl". 

"hover" -> when you hover over the variable from the editor.
"clipboard" -> "Copy Value" context menu from VARIABLES or WATCH view

**Comment by hyangah** [comment]
PR description is updated. Seems like we all agreed on 256 for now.
@suzmue has a prototype of removing type info and slice/array's len/cap info from the value presentation and making them available in a different way (type info is already supplied as the separate type attribute)

**Comment by polinasok** [comment]
> "hover" -> when you hover over the variable from the editor.

I believe this is only true when you hover over the code, not in the Variables section.

> @suzmue has a prototype of removing type info and slice/array's len/cap info from the value presentation 
Should we consider smaller limit once the type info is dropped from the inlined value?

**Comment by polinasok** [comment]
@hyangah Oh and sorry for editing the comment instead of replying.

**Comment by hyangah** [comment]
>> "hover" -> when you hover over the variable from the editor.
>
> I believe this is only true when you hover over the code, not in the Variables section.

fyi @polinasok 'editor' refers to the main area where code/document is presented in vscode. 'views' or 'viewlets' are the sections presented in the side bar area.

**Pull request merged**